src/main/java/org/radarcns/service/DataSetService.java

```diff
@@ -306,7 +292,13 @@ private DataItem collectDistinctSources(String projectName, String subjectId,
             TimeWindow timeWindow) {
         int count = aggregateDataSources.stream()
                 .map(aggregate -> (int) aggregate.getSourceData().stream()
-                        .map(sourceData -> mongoSensorMap.get(sourceData.getName()))
+                        .map(sourceData -> {
+                            try {
+                                return this.sourceCatalog.getSourceData(sourceData.getName());
+                            } catch (IOException exe) {
+                                throw new BadGatewayException(exe);
```

Comment:
Why are we using BadGatewayException ?

Reply:
Because the sourceData cannot be successfully retrieved from MP, where `rest-api` is the Gateway for the client requesting. It is a custom runtime exception, which an ExceptionMapper to respond with proper HTTP status.